docs(dev): container execution backend + ricochet-server in a container (#237)
ricochet-rs/docs: src/content/docs/dev/admin/configuration/execution-backend.mdx

import { Aside } from "@astrojs/starlight/components";

> ricochet runs user content — apps and tasks — in an **execution backend**.

Reviewer: ```suggestion
ricochet runs content items in an **execution backend**.
```

Three backends are supported:

| Backend      | `engine` value | Use when                                                                                                          |
| ------------ | -------------- | ----------------------------------------------------------------------------------------------------------------- |

> | Host         | (omitted)      | Single host, no container daemon. Content runs as isolated processes on the host via [libcontainer][libcontainer]. |

Reviewer: ```suggestion
| Host         | (omitted)      | Content runs as a rootless OCI container. |
```

> | Container    | `container`    | Single host with a Docker- or Podman-compatible daemon. Content runs as full containers spawned through the daemon. |

Reviewer: ```suggestion
| Container    | `container`    | Single host with a Docker or Podman-compatible daemon. Content runs as a container spawned through the daemon. |
```

| Kubernetes   | `k8s`          | Multi-node cluster, content scheduled as Pods.                                                                    |

> [libcontainer]: https://docs.rs/libcontainer/

Reviewer: ```suggestion
```

If `[backend]` is omitted, ricochet uses the **host** backend.

> Content runs as isolated processes managed in-process by `libcontainer` (an OCI-runtime implementation, no daemon required), using the language interpreters installed on the host.

Reviewer: ```suggestion
Content runs a rootless OCI-compatible container using the language interpreters installed on the host.
```

> This page documents the `container` and `k8s` backends; the host backend has no configuration knobs of its own.

Reviewer: ```suggestion
```

semi-colons—the new AI tic.

<Aside type="tip" title="Default exec-env images">

>   The container and Kubernetes backends pull an OCI image per content item.

Reviewer: ```suggestion
```
I don't think this is necessary

>   ricochet publishes a continuously-updated set of production-ready images at [ricochet-rs/exec-envs](https://github.com/ricochet-rs/exec-envs) — R on Alpine, Ubuntu, and AlmaLinux; Python on Ubuntu and AlmaLinux — rebuilt weekly and suitable as defaults for `default_image` and per-content `exec_env` references.

Reviewer: ```suggestion
  ricochet publishes a continuously-updated set of production-ready images at [ricochet-rs/exec-envs](https://github.com/ricochet-rs/exec-envs) which are rebuilt weekly and are suitable defaults.
```

  Bring your own images when you need additional system libraries or a different language/version combination.
</Aside>

## Host vs container on a single host

Both backends run content on the same machine as ricochet-server.
The difference is **how** content is isolated and **where** the runtime environment comes from.

> | Dimension                | Host backend (`libcontainer`)                                          | Container backend (`container`)                                  |

Reviewer: ```suggestion
| Dimension                | Host backend                                    | Container backend (`container`)                                  |
```

| ------------------------ | ---------------------------------------------------------------------- | ---------------------------------------------------------------- |

> | Isolation mechanism      | `libcontainer` namespaces/cgroups, set up in-process by ricochet-server | OCI containers, spawned by the host's Docker/Podman daemon       |

Reviewer: ```suggestion
| Isolation mechanism      | Namespace and cgroup with PID isolation | OCI containers, spawned by the host's Docker/Podman daemon       |
```

| Daemon required          | No                                                                     | Yes (Docker, Podman, or compatible)                              |
| Runtime / interpreters   | Whatever is installed on the host (system R, Python, Julia)            | Whatever the chosen image provides; arbitrary per content item   |

> | Per-content exec env     | Single shared host environment                                         | Per-item via `exec_env` → image, like the Kubernetes backend     |

Reviewer: These seem redundant. The "per-content exec-env" row can probably be deleted?

> | Per-content exec env     | Single shared host environment                                         | Per-item via `exec_env` → image, like the Kubernetes backend     |

Reviewer: ```suggestion
| Interpreters   | Discovered on the host            | Defined by the execution environment   |
```

| Observability            | Processes are not visible to `docker ps` / container metrics tooling   | Each app/task is a real container — visible to standard tooling  |
| Setup cost               | None — works on a fresh ricochet-server install                        | Install and run a container daemon, grant socket access          |

> Pick the **host** backend for simple single-tenant deployments where the host already has the interpreters you need and you do not want to operate a container daemon.

Reviewer: ```suggestion
Use the **host** backend for deployments where the host already has the interpreters you need and you do not want to operate a container daemon.
```

> Pick the **container** backend when you need different runtime environments per content item (e.g. R 4.3 vs R 4.5, conflicting system libraries), or when you want each running app and task to be a visible, individually observable container.

Reviewer: ```suggestion
Use the **container** backend when you want each running app and task to be a visible in an observable container.
```

Different runtimes are already supported.

## Container backend

Set `engine = "container"` to spawn content as containers via a local Docker- or Podman-compatible daemon.

```toml title="ricochet-config.toml"
[backend]
engine = "container"
default_image = "r-default"
exec_env_config_path = "ricochet-exec-env.toml"
content_path = "/var/lib/ricochet/data/content"
cache_path = "/var/lib/ricochet/data/.cache"
```

| Field                  | Required | Description                                                                                                                                                  |
| ---------------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `engine`               | yes      | Must be `"container"`.                                                                                                                                       |

> | `default_image`        | no       | Image category name from `exec_env_config_path`. Used when a content item does not specify its own exec environment.                                         |

Reviewer: This "category" language needs to be removed 😭 
```suggestion
| `default_image`        | no       | An image name from the `ricochet-exec-env.toml` to be used as a default when no `exec-env` is specified.                                         |
```

| `exec_env_config_path` | yes      | Path to the exec-env config TOML. Absolute, or relative to `RICOCHET_HOME`.                                                                                  |

> | `content_path`         | yes      | Host path bind-mounted into every spawned container at `/opt/ricochet/data/content`. Must contain ricochet's deployment bundles.                             |

Reviewer: ```suggestion
| `content_path`         | yes      | Path that is bind-mounted into every spawned container. Must contain ricochet's deployment bundles.                             |
```

> | `cache_path`           | yes      | Host path bind-mounted into every spawned container at `/opt/ricochet/cache`. Stores R, Julia, and Python package caches shared across content.              |

Reviewer: ```suggestion
| `cache_path`           | yes      | Path bind-mounted into every spawned container. Stores R, Julia, and Python package caches shared across content.              |
```

> | `proxy_host`           | no       | Hostname ricochet's reverse proxy uses to reach spawned content containers. Defaults to `localhost`; override when running ricochet-server in a container (see [Deployment scenarios](#deployment-scenarios)). |

Reviewer: ```suggestion
| `proxy_host`           | no       | Hostname ricochet's reverse proxy uses to reach spawned content containers. Defaults to `localhost`. Override when running ricochet in a container (see [Deployment scenarios](#deployment-scenarios)). |
```

ricochet refuses to start if `exec_env_config_path`, `content_path`, or `cache_path` is missing.

### Image resolution

`default_image` and per-content `exec_env` values are looked up in the exec-env config.

> A category like `r-default` resolves to a concrete image reference (`docker.io/ricochetrs/r-alpine:4.5`).

Reviewer: ```suggestion
For example, `r-default` resolves to a concrete image reference (`docker.io/ricochetrs/r-alpine:4.5`).
```
